packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java

```diff
@@ -337,15 +338,15 @@ private static Layout createLayout(
       int hintWidth = (int) Math.ceil(desiredWidth);
       layout =
           StaticLayout.Builder.obtain(text, 0, spanLength, sTextPaintInstance, hintWidth)
-              .setAlignment(Layout.Alignment.ALIGN_NORMAL)
+              .setAlignment(alignment)
               .setLineSpacing(0.f, 1.f)
               .setIncludePad(includeFontPadding)
               .setBreakStrategy(textBreakStrategy)
               .setHyphenationFrequency(hyphenationFrequency)
               .build();
 
     } else if (boring != null && (unconstrainedWidth || boring.width <= width)) {
-      int boringLayoutWidth = boring.width;
+      int boringLayoutWidth = (int) width;
```

Comment:
I feel like this should be safe to change since the boring layout should be one-line anyway and making it fill the entire width is necessary to make different alignments actually work. Otherwise, it looks like it's trying to align the text in a vies just big enough to fit it so it basically doesn't do anything.

Reply:
On web, if a width isn't set, we don't expand. https://jsfiddle.net/egnxrs24/1/

What does `width` here represent? Available space? I think the ideal would be that we don't expand the intrinsic size for single line text, but that width could be explicitly set.

Reply:
When I tested browser behavior, alignment on paragraph element (i.e. top-level text span) will account for all the space: https://jsfiddle.net/z02a1xdh/1/ When set on span, it does not expand.

We should make sure to not regress https://github.com/facebook/react-native/pull/37590

Reply:
You're right, I've updated this change (and added an analogous one in the first case) to use `width` only when `widthYogaMeasureMode` is set to `EXACTLY`, which, as long as the name of the mode isn't deceptive, is the case where the text should be exactly that width.

<details>
<summary>As for the linked issue I've tested it on the following code</summary>

```jsx
      <Text>just 1</Text>
      <Text style={{backgroundColor: 'red', alignSelf: align, color: 'white'}}>
        1
      </Text>
      <Text>1 + {"{new line}:"}</Text>
      <Text style={{backgroundColor: 'red', alignSelf: align, color: 'white'}}>
        1{'\n'}
      </Text>
      <Text>1 + {"{bunch of spaces}:"}</Text>
      <Text style={{backgroundColor: 'red', alignSelf: align, color: 'white'}}>
        1{'                                                                              '}
      </Text>
      <Text>1 + {"{bunch of spaces + new line}:"}</Text>
      <Text style={{backgroundColor: 'red', alignSelf: align, color: 'white'}}>
        1{'                                                                              \n'}
      </Text>
      <Text>1 + {"{bunch of spaces (overflowing)}:"}</Text>
      <Text style={{backgroundColor: 'red', alignSelf: align, color: 'white'}}>
        1{'                                                                                                              '}
      </Text>
      <Text>1 + {"{bunch of spaces (overflowing) + something}:"}</Text>
      <Text style={{backgroundColor: 'red', alignSelf: align, color: 'white'}}>
        1{'                                                                                                              x'}
      </Text>
      <Text>1 + {"{bunch of spaces (overflowing) + new line}:"}</Text>
      <Text style={{backgroundColor: 'red', alignSelf: align, color: 'white'}}>
        1{'                                                                                                              x\n'}
      </Text>

      <Text>1 + bunch of spaces (overflowing)</Text>
      <Text style={{backgroundColor: 'red', alignSelf: align, color: 'white'}}>
        1                                                                                                              
      </Text>
      <Text>1 + bunch of spaces (overflowing) + something</Text>
      <Text style={{backgroundColor: 'red', alignSelf: align, color: 'white'}}>
        1                                                                                                      
        x
      </Text>
      <Text>1 + bunch of spaces (overflowing) + something + new line</Text>
      <Text style={{backgroundColor: 'red', alignSelf: align, color: 'white'}}>
        1                                                                                                              x{'\n'}
      </Text>
```

</details>

And this is the result I got:

|iOS new arch|Android old arch|Android new arch without changes|Android new arch with changes|
|-|-|-|-|
|<img width="502" alt="ios_new" src="https://github.com/facebook/react-native/assets/21055725/934c04bc-944e-4baa-afc3-53bce4b4e604">|<img width="411" alt="android_old" src="https://github.com/facebook/react-native/assets/21055725/baec52c3-0eb0-4db8-8590-e35734ad36da">|<img width="411" alt="android_new_no_changes" src="https://github.com/facebook/react-native/assets/21055725/e4f28f1c-d3b9-4cce-9b82-c3502cd321a9">|<img width="411" alt="android_new_changed" src="https://github.com/facebook/react-native/assets/21055725/5f787bf8-83c9-4baf-a8a6-1e662baf0c26">|

Let me know if I should try different cases.

Reply:
I also changed simple casts from `float` to `int` to take the ceiling, as it was producing weird layout issues where the width was set to `x.999` and the decimal part was discarded.

Reply:
|Without `Math.ceil`|With `Math.ceil`|
|-|-|
|<img width="411" alt="Screenshot 2024-04-22 at 15 49 55" src="https://github.com/facebook/react-native/assets/21055725/7b8fa6ea-8555-4055-81f1-b98a7936209e">|<img width="411" alt="Screenshot 2024-04-22 at 15 49 42" src="https://github.com/facebook/react-native/assets/21055725/3ec98991-8b00-40ee-9d62-29f668a7b9ea">|

```jsx
<Text style={{backgroundColor: 'red', alignSelf: 'flex-end', color: 'white', textAlign: 'right'}}>
  AB<View style={{width: 25, height: 20, backgroundColor: 'blue'}} />CD
</Text>
```